core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -361,7 +361,7 @@ object Uri {
         val (start, end) = segments.splitAt(idx)
         Path(start, absolute = absolute) -> Path(end, true, endsWithSlash = endsWithSlash)
       }
-    private def copy(
+    def copy(
```

Comment:
Are there any invariants we might be breaking because of making this public?  Certain invalid combinations we don't want to allow?  I _think_ all states are legal.

Reply:
@rossabaker really sorry, I did not understand why/how does variance come into the picture here?

Reply:
Invariants (e.g., constraints), not invariance (e.g., subtypes and supertypes).  Invariants are properties of a class that can't be violated.  Silly example:

```scala
class Date(month: Int, day: Int, year: Int)
```

Invariants of `Date` might be that `month >= 1` and `month <= 12` and `if month == 4 then date <= 30`.  When we expose a copy constructor, we want to e sure that all parameters to it are legal, and all combinations are legal.  And whenever a constructor is private, it's often the case that `copy` should be private.

Reply:
@rossabaker ah okay, got it. Makes sense. I used the copy method purely for convenience. Should I revert it?

Reply:
Let's see what @hamnis says.  If it's dangerous, we can revert and add appropriate `with*` methods.  If it's safe, we can keep the convenience.  Either way, we should have a clear path forward.